### PR TITLE
Create Add, Remove, and Find CD entrypoints for scripting API

### PR DESCRIPTION
**Pull request by Sarci1**
- [X ] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/niveristand-scan-engine-ethercat-custom-device/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

Creates Add, Find, and Remove Scan Engine CD methods for the scripting API. Also addresses Item Reference To Pointer.vi calls in dependent subVIs.

### Why should this Pull Request be merged?

Beginning of a scripting API for the Scan Engine CD.

### What testing has been done?

Manual testing and validation of behavior for adding and removing the custom device.

### CODE REVIEW

**Review by rtzoeller — COMMENTED** [review]
Some of my feedback is similar to feedback I provided @dbendele on https://github.com/ni/niveristand-data-sharing-framework-custom-device/pull/19.

- [x] `Add Scan Engine Custom Device.vi`: I would love to see proper error codes used here. The R+F API used stub error codes as well (single digit negative integers), but we never tracked the work to replace them. 

image

The error codes file for this custom device is at `Custom Device Source/Docs/scanengineveristand-errors.txt`. I believe we only install it to the LabVIEW Runtime location currently, so when we start packaging the scripting API we will also need to install it to the dev. location.

- [x] `Find Scan Engine Custom Device.vi`: This is another place where R+F does a poor job that we should deviate from (like the DSF PR, this PR is a running list of things to go clean up in R+F 🤦‍♂️). The VeriStand API defines `0xFFFFFFFFFFFFFFFF` as the invalid node reference, not zero (see `NI VeriStand - Not a Ref.vi`).
I'd prefer to see this VI return that value if the custom device is not found, and also to have it return a boolean indicating if the custom device is found (like `NI VeriStand - Search for Item.vi`).

- [x] `Validate Reference by Name.vi`: This is another place where we are returning an generic error code.

- [x] `main settings.ctl`: I am seeing a broken typedef. Also having issues with `Set Main Settings.vi`. Did the library get saved properly?

- [x] `NI VeriStand - Get Target Operating System`: ~~Can this VI be removed and replace with vi.lib's `NI VeriStand - Get Target Type`?~~ No, because VeriStand apparently never added support for Linux to that VI for the offline case. Cool. Bug 1134478 created.

**Comment by rtzoeller** [comment]
@dbendele adding you as a reviewer, because the API is very similar to your DSF work and I'd love to see alignment between them (since R+F has set a poor example).

**Comment by rtzoeller** [comment]
@Sarci1 your build here is going to keep failing because VxWorks support no longer builds against 2020 R3 (independent of your change). I'll get a fix in shortly so your PR build can succeed.

**Comment by Sarci1** [comment]
@rtzoeller regarding the error codes, who do I work with to get an appropriate chunk of numbers for this effort? I imagine it needs to come from the pre-allotted VeriStand chunk.

**Comment by rtzoeller** [comment]
> @rtzoeller regarding the error codes, who do I work with to get an appropriate chunk of numbers for this effort? I imagine it needs to come from the pre-allotted VeriStand chunk.

Unless @niphilj disagrees, I would prefer if the error codes were sequential with the existing ones for the custom device (see [here](https://github.com/ni/niveristand-scan-engine-ethercat-custom-device/blob/main/Custom%20Device%20Source/Docs/scanengineveristand-errors.txt)). That is in the "user defined" error codes range, but I don't think requesting a new range and moving the existing error codes is in scope for this work.

Yielding to @niphilj.

**Comment by Sarci1** [comment]
> * [x]  `Add Scan Engine Custom Device.vi`: I would love to see proper error codes used here. The R+F API used stub error codes as well (single digit negative integers), but we never tracked the work to replace them.
>
Done. Updated errors.txt and Error Strings.vi.

> * [x]  `Find Scan Engine Custom Device.vi`: This is another place where R+F does a poor job that we should deviate from (like the DSF PR, this PR is a running list of things to go clean up in R+F 🤦‍♂️). The VeriStand API defines `0xFFFFFFFFFFFFFFFF` as the invalid node reference, not zero (see `NI VeriStand - Not a Ref.vi`).
>
Done. Had to create a new global because other was not accessible in this context.

>   I'd prefer to see this VI return that value if the custom device is not found, and also to have it return a boolean indicating if the custom device is found (like `NI VeriStand - Search for Item.vi`).
>
Added boolean "found."

> * [x]  `Validate Reference by Name.vi`: This is another place where we are returning an generic error code.
>
Done. Updated errors.txt and Error Strings.vi.

> * [x]  `main settings.ctl`: I am seeing a broken typedef. Also having issues with `Set Main Settings.vi`. Did the library get saved properly?
>
Should be fixed. I didn't save it properly.

> * [x]  `NI VeriStand - Get Target Operating System`: ~Can this VI be removed and replace with vi.lib's `NI VeriStand - Get Target Type`?~ No, because VeriStand apparently never added support for Linux to that VI for the offline case. Cool. Bug 1134478 created.
>
Completed by @rtzoeller

**Comment by rtzoeller** [comment]
@Sarci1 you have a conflict with `main` for the file `Custom Device Source/Scan Engine.lvproj`. It should be straightforward to merge `main` into your branch in order to resolve the conflict; all the other commit did was delete the VxWorks section of the project.

**Review by rtzoeller — APPROVED** [review]
Final thoughts:

- [ ] `Validate Reference by Name.vi` should probably be replaced by a `Validate Reference by Guid.vi`. I can rename an Embedded Data Logger custom device to `Scan Engine and EtherCAT` - that doesn't make it an instance of this custom device. Another place where R+F got it wrong (although I think it uses both).

- [ ] It seems like the scripting API is not properly associated with a library.
image

**Comment by Sarci1** [comment]
I'll make those changes in another pull request. Thank you!

**Comment by Sarci1** [comment]
* [x]  `Validate Reference by Name.vi` should probably be replaced by a `Validate Reference by Guid.vi`. I can rename an Embedded Data Logger custom device to `Scan Engine and EtherCAT` - that doesn't make it an instance of this custom device. Another place where R+F got it wrong (although I think it uses both).

Done and I updated the Remove function to use it for both GUIDs (enabled/disabled). I also created a new error code for this function. I left the Validate by Name for now in case I need to use it in the future for some reason.

* [x]  It seems like the scripting API is not properly associated with a library.
  image

This was dumb on my part... fixed it.

**Comment by niphilj** [comment]
> Unless @niphilj disagrees, I would prefer if the error codes were sequential with the existing ones for the custom device
I don't disagree.

**Review by niphilj — APPROVED** [review]

**Comment by niphilj** [comment]
> @dbendele adding you as a reviewer, because the API is very similar to your DSF work and I'd love to see alignment between them (since R+F has set a poor example).

I'm assuming this is more of an FYI aimed at Doug.

File: Custom Device Source/Docs/scanengineveristand-errors.txt [hunk 1]
```diff
@@ -30,4 +30,20 @@ Unsupported file type.  Please select a valid .lvbitx file.
...
+<nierror code="537714">
+Timeout communicating with EtherCAT reconfiguration loop.
+</nierror>
+<nierror code="537715">
+The Scan Engine and EtherCAT custom device already exists.
+</nierror>
+<nierror code="537716">
+The reference ID provided does not match the expected name or cannot be found.
+</nierror>
+</nierror>
```
Comment (niphilj) [inline Custom Device Source/Docs/scanengineveristand-errors.txt:45]: Is there supposed to be 2 close tags?

Comment (niphilj) [reply Custom Device Source/Docs/scanengineveristand-errors.txt:45]: I'll merge it after a confirmation. Thanks.

Comment (Sarci1) [reply Custom Device Source/Docs/scanengineveristand-errors.txt:45]: @niphilj nope, you are correct. That's a mistake. I'll correct it.

Comment (Sarci1) [reply Custom Device Source/Docs/scanengineveristand-errors.txt]: @niphilj I've committed it.

**Review by buckd — APPROVED** [review]

**Review by dbendele — APPROVED** [review]
My own personal preference is to use individual elements (**sync to scan?**, **scan engine decimation**, **delay after FPGA load (s)**) rather than cluster (**settings**). Given the expected stability of this API (and my poor timing), NO change requested.
***
- [ ] Do not use "?" for Boolean inputs that are implied commands such as **sync to scan?** in Add Scan Engine Custom Device.vi
   reference: https://dev.azure.com/ni/DevCentral/_wiki/wikis/AppCentral.wiki/15392/LabVIEW-NXG-Style-Guide?anchor=12-api-interface-(icon-and-connector-pane)
   Note: the "?" is appropriate for implied questions as in **sync to scan?** output of Is Synced to Scan.vi
- [ ] control and indicator names should be lower case Invalid Ref --> invalid reference, Root Item Ref --> root item reference
- [ ] control and indicator names that are acronyms should follow standard capitalization guid --> GUID

**Pull request merged**